### PR TITLE
add k8s.cluster.uid resource attribute semconv

### LINKED ISSUE #3266: Proposal for `k8s.cluster.uid` resource attribute
**What are you trying to achieve?**

Kubernetes doesn't have a formal way to identify a cluster, but a pseudo-id can be derived by taking the `uid` of the `kube-system` namespace. I'd like to propose that we add a semantic convention `k8s.cluster.uid` set to the `uid` of the `kube-system` namespace. Additionally I'll open a PR with the proposed changes and we can discuss pros and cons and alternative approaches here or there.

### PR DESCRIPTION
**Pull request by mwear**
Fixes #3266

## Changes

Kubernetes doesn't have a formal way to identify a cluster, but a pseudo-id can be derived by taking the `uid` of the `kube-system` namespace. This PR adds a convention for `k8s.cluster.uid` to be set to the `kube-system` namespace uid.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
> Kubernetes doesn't have a formal way to identify a cluster, but a pseudo-id can be derived by taking the `uid` of the `kube-system` namespace.

Any links that describe that this will work as expected? (e.g. is persisted, globally unique, etc).

**Comment by mwear** [comment]
Here is a link to the kubernetes documentation for object UIDs: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids.

The main point is:

> Every object created over the whole lifetime of a Kubernetes cluster has a distinct UID. It is intended to distinguish between historical occurrences of similar entities.

My understanding is that the kube-system namespace should exist for the lifetime of the cluster and would only change if the cluster was recreated. One slight caveat, is that UIDs are unique within a cluster, but there is no uniqueness guarantee between clusters. It is possible, although very unlikely, that two separate clusters could independently generate the same UID for the kube-system namespace.

**Comment by mwear** [comment]
A little more on the issue of uniqueness of UIDs between clusters. The same [k8s doc](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids) adds:

> Kubernetes UIDs are universally unique identifiers (also known as UUIDs). UUIDs are standardized as ISO/IEC 9834-8 and as ITU-T X.667.

I found the following from the [ITU](https://www.itu.int/en/ITU-T/asn1/Pages/UUID/uuids.aspx):

> If generated according to one of the mechanisms defined in Rec. [ITU-T X.667 | ISO/IEC 9834-8](https://www.itu.int/ITU-T/studygroups/com17/oid.html), a UUID is either guaranteed to be different from all other UUIDs generated before 3603 A.D., or is extremely likely to be different (depending on the mechanism chosen). The UUID generation algorithm specified in this standard supports very high allocation rates: 10 million per second per machine if necessary, so UUIDs can also be used as transaction IDs

I think we can say that it is "extremely unlikely" that a UID will conflict between clusters.

**Review by carlosalberto — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@dmitryax can you please take a look? I think it is very useful to have a unique cluster identifier.

**Comment by carlosalberto** [comment]
@dmitryax any opinion? ;)

**Comment by dmitryax** [comment]
The source of the data seems a bit hacky to me, but, I agree, it is the most reliable option available now. My concern is, if a more appropriate source becomes available in the future, how can we migrate to it?

@mwear I'm curious what are your plans on utilizing this attribute?

File: specification/resource/semantic_conventions/k8s.md [hunk 1]
```diff
@@ -25,8 +25,16 @@ Kubernetes object, but "name" is usually more user friendly so can be also set.
 | Attribute  | Type | Description  | Examples  | Requirement Level |
 |---|---|---|---|---|
 | `k8s.cluster.name` | string | The name of the cluster. | `opentelemetry-cluster` | Recommended |
+| `k8s.cluster.uid` | string | A pseudo-id for the cluster, set to the uid of the `kube-system` namespace. | `218fc5a9-a5f1-4b54-aa05-46717d0ab26d` | Recommended |
 <!-- endsemconv -->
 
+**Note on k8s.cluster.id:**
```
Comment (joaopgrassi) [inline specification/resource/semantic_conventions/k8s.md:31]: I think this text needs to also be part of the yaml definition. You can use a `note` in there and then this is automatically "linked" at the bottom. You can see an example here https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/messaging.yaml#L45

**Comment by joaopgrassi** [comment]
>My concern is, if a more appropriate source becomes available in the future, how can we migrate to it?

Since the text is "just a recommendation" and nothing normative, I think it can change later to indicate the newer, better approach without causing much issues or?

**Comment by mwear** [comment]
This would be a more reliable way to correlate data to a k8s cluster than using name alone. It would be ideal if this became something supported by the k8s API in the future, but this is the best we have for now. If the source were to change in the future, we should be able to handle this via a telemetry schema. If I'm mistaken on this point, let me know and we could consider a different name for this attribute.

File: specification/resource/semantic_conventions/k8s.md (hunk 1 above)
Comment (mwear) [reply specification/resource/semantic_conventions/k8s.md]: thanks! i added this as a note to the yaml def.

**Comment by dmitryax** [comment]
Sounds good to me

**Comment by dmitryax** [comment]
> If the source were to change in the future, we should be able to handle this via a telemetry schema. If I'm mistaken on this point, let me know and we could consider a different name for this attribute.

Not sure how this can be approached via telemetry schema keeping the same attribute name. And I don't have any other name suggestions for this attribute, `k8s.cluster.uid` looks good to me. We could consider designating the current source as a recommendation and indicating in the notes that it is subject to change in the future, as k8s may yield more suitable sources, WDYT?

**Review by dmitryax — APPROVED** [review]

**Review by tigrannajaryan — APPROVED** [review]

File: semantic_conventions/resource/k8s.yaml [hunk 2]
```diff
@@ -10,6 +10,35 @@ groups:
         brief: >
           The name of the cluster.
         examples: ['opentelemetry-cluster']
+      - id: uid
+        type: string
+        brief: >
+          A pseudo-id for the cluster, set to the uid of the `kube-system`
```
Comment (MrAlias) [inline semantic_conventions/resource/k8s.yaml:16]:
```suggestion
          A pseudo-ID for the cluster, set to the UID of the `kube-system`
```

File: semantic_conventions/resource/k8s.yaml [hunk 3]
```diff
@@ -10,6 +10,35 @@ groups:
...
+          namespace.
+        note: |
+          K8s does not have support for obtaining a cluster id. If this is ever
```
Comment (MrAlias) [inline semantic_conventions/resource/k8s.yaml:19]:
```suggestion
          K8s does not have support for obtaining a cluster ID. If this is ever
```

File: semantic_conventions/resource/k8s.yaml [hunk 4]
```diff
@@ -10,6 +10,35 @@ groups:
...
+          added, we will recommend collecting the `k8s.cluster.uid` through the
+          official APIs. In the meantime, we are able to use the `uid` of the
+          `kube-system` namespace as a proxy for cluster id. Read on for the
```
Comment (MrAlias) [inline semantic_conventions/resource/k8s.yaml:22]:
```suggestion
          `kube-system` namespace as a proxy for cluster ID. Read on for the
```

**Review by MrAlias — APPROVED** [review]

**Comment by carlosalberto** [comment]
@joaopgrassi @arminru Last chance to review ;)

**Review by arminru — APPROVED** [review]

**Review by joaopgrassi — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

**Pull request merged**